rgw multisite: bucket reshard work in progress

cbodley opened the pull request:
stable working branch for bucket resharding support in multisite

design document: https://github.com/ceph/ceph/blob/master/src/doc/rgw/multisite-reshard.md

## todo
- [x] store index layout and list of log layouts in bucket info (merged to master+pacific)
- [x] fix initialization of index layout and 'logs' for master+pacific (merged in https://github.com/ceph/ceph/pull/39032)
- [x] bucket-wide full sync and sync status (merged in https://github.com/ceph/ceph/pull/37573)
- [x] reshard modifies existing bucket instance instead of creating a new one (merged in https://github.com/ceph/ceph/pull/38657, was https://github.com/ceph/ceph/pull/35175)
- [x] reshard preserves old index objects if multisite (merged in https://github.com/ceph/ceph/pull/40446)
- [x] extend bilog admin APIs with log layout generation information (merged https://github.com/ceph/ceph/pull/38586)
- [x] data sync tracks bucket generation numbers in datalog, error repo, ~~notifications~~ (merged in https://github.com/ceph/ceph/pull/39396, was https://github.com/ceph/ceph/pull/38561)
- [x] data sync tracks bucket generation numbers in async notifications (merged in https://github.com/ceph/ceph/pull/39477)
- [x] incremental bucket sync updates bucket-wide sync status when a shard finishes the current log generation (merged in #40605 @smanjara)
- [x] clean up shard status objects after finishing incremental sync on a generation (https://github.com/ceph/ceph/pull/42459)
- [x] as the first step for the enable disable, we need this fix (merged in https://github.com/ceph/ceph/pull/41337)
- [x] bucket sync disable/enable at bucket level (merged in https://github.com/ceph/ceph/pull/41404)
- [x] json formatter for BucketLayout structure (merged in https://github.com/ceph/ceph/pull/39231)
- [x] bilog trimming per log generation (merged in https://github.com/ceph/ceph/pull/41051 @adamemerson)
- [ ] test cases for bilog trim across reshard events (https://github.com/ceph/ceph/pull/44758 @TRYTOBE8TME)
- [x] zonegroup flag to enable resharding, along with radosgw-admin command and documentation (merged https://github.com/ceph/ceph/pull/39927)
- [x] test cases for bucket sync (both full and incremental) across reshard events (merged https://github.com/ceph/ceph/pull/41110)
- [x] teach 'bucket sync checkpoint' about generations for testing (merged https://github.com/ceph/ceph/pull/41455)
- [x] test suite for sync between zones of different ceph versions  (merged https://github.com/ceph/ceph/pull/39456)
- [x] on bucket reshard, write datalog entries for each shard of the new generation (merged #41471)
- [x] track shard sync status objects per generation (merged https://github.com/ceph/ceph/pull/41538)
- [x] audit sync's use of num_shards from bucket info - most shard counts should come from remote via BILog_Info/List (wip https://github.com/ceph/ceph/pull/41810 @smanjara)
- [ ] add tests that combine reshard with enable/disable (https://github.com/ceph/ceph/pull/44838 @TRYTOBE8TME)
- [x] limit number of bilog generations; prevent reshard (dynamic and manual) if it's at the max, never trim if there's only one (merged https://github.com/ceph/ceph/pull/43334)
- [x] commit_reshard() and cancel_reshard() should keep retrying on ECANCELED (merged #44990)
- [x] preserve xattrs like RGW_ATTR_ACL and RGW_ATTR_LC over commit_reshard()/cancel_reshard() (merged https://github.com/ceph/ceph/pull/44788)
- [x] the 'bucket sync cache' ignores generations when comparing timestamps, so may discard entries with newer generation (merged https://github.com/ceph/ceph/pull/45122/)
- [x] resolve "WARNING: The bucket info cache is inconsistent" warnings from reshard (merged #44853)
- [x] bilog trimming shouldn't erase xattrs on the bucket instance (merged #44871)
- [ ] check for reshard status via `fetch_new_bucket_info` before calling `wait()` in `RGWRados::block_while_resharding()`. after the fix clients should not be blocked for 10s during reshard
- [x] rewrite `radosgw-admin bucket sync run` to transition over generations
---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mattbenjamin commented:
@adam @cbodley this PR seems like it needs an item in the backlog?  https://github.com/ceph/ceph/pull/38230

cbodley commented:
> @adam @cbodley this PR seems like it needs an item in the backlog? #38230

@mattbenjamin this pr is only tracking the multisite resharding work. @adamemerson's 'FIFO the Datalog' work merged for pacific, so i'd rather we track the related backports as tracker issues and BZs

mattbenjamin commented:
@cbodley @smanjara @adamemerson  is it possible we could update the still open items w/owners where an owner exists?

adamemerson commented on src/rgw/rgw_reshard.cc:
```diff
@@ -285,174 +272,303 @@ int RGWBucketReshard::set_resharding_status(const DoutPrefixProvider *dpp,
-      return ret;
+  if (!store->svc()->zone->need_to_log_data()) {
+    // if we're not syncing data, we can drop any existing logs
+    layout.logs.clear();
```
@cbodley Is this safe? We use layout.logs.back() in enough places that wiping out all logs doesn't seem like something we can get away with.

cbodley replied:
accessing logs.back() outside of multisite code does not seem safe, no

i think this reshard code is correct, because outside of multisite, we'll always delete the previous generation's index objects. so logs.back() would point at missing objects if we left it in the vector

as @ivancich found in https://github.com/ceph/ceph/pull/44393, indexless buckets are another case where we wouldn't have any logs

cbodley replied:
er, this is in `commit_reshard()`, which goes on to add the *new* index layout to `logs` with:
```c++
layout.logs.push_back(log_layout_from_index(next_log_gen, layout.current_index));
```
so even with multisite disabled, we'll still be able to rely on `logs.back()`

cbodley commented:
@TRYTOBE8TME to resolve the `Health check failed: Module 'feedback' has failed: Not found or unloadable (MGR_MODULE_ERROR)` failures, you'll need to pull in the 2 commits from https://github.com/ceph/ceph/pull/44685

cbodley commented:
some rgw_sal_motr-related failures crept in with the rebase, but it's not a big deal until we're ready to merge this

cbodley commented:
it looks like the json encoding of zone/zonegroup features was accidentally removed when rebasing over https://github.com/ceph/ceph/pull/44299. i'll raise a PR to add them back (cc @yuvalif)

vumrao commented:
@vumrao

cbodley commented:
getting this ready to merge. there's a compilation failure in test_rgw_bucket_sync_cache.cc, fixed in https://github.com/ceph/ceph/pull/46290

cbodley commented:
seeing teuthology failures on `FAIL: s3tests_boto3.functional.test_s3.test_set_bucket_tagging`, so i rebased over current main to pick up https://github.com/ceph/ceph/pull/46167. rebased without conflict

adamemerson commented:
Jenkins, test API

cbodley commented:
failed qa in https://pulpito.ceph.com/cbodley-2022-05-27_19:16:02-rgw-wip-rgw-multisite-reshard-distro-default-smithi/

cbodley commented:
not sure why this multisite test is failing:
```
2022-06-01T03:13:06.134 DEBUG:teuthology.orchestra.run.smithi063:> adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage radosgw-admin --log-to-stderr --format json -n client.0 --cluster c1 bucket reshard --bucket qmhdzr-49 --num-shards 3 --yes-i-really-mean-it --cluster c1 --debug-rgw 0 --debug-ms 0
2022-06-01T03:13:06.190 INFO:teuthology.orchestra.run.smithi063.stderr:ignoring --setuser ceph since I am not root
2022-06-01T03:13:06.206 INFO:teuthology.orchestra.run.smithi063.stderr:ignoring --setgroup ceph since I am not root
2022-06-01T03:13:06.320 INFO:teuthology.orchestra.run.smithi063.stderr:The zonegroup 'a' does not have the resharding feature enabled.
2022-06-01T03:13:06.322 DEBUG:teuthology.orchestra.run:got remote process result: 95
2022-06-01T03:13:06.400 INFO:tasks.rgw_multisite_tests:rgw_multi.tests.test_bucket_reshard_index_log_trim ... FAIL
```

cbodley commented:
mostly passed qa in https://pulpito.ceph.com/cbodley-2022-06-01_19:25:46-rgw-master-distro-default-smithi/

mattbenjamin reviewed (approved):
we are not worthy!

yuvalif commented:
this deserves a celebration 🥳 🎉
